// ngx-material-timepicker/esm5/src/app/material-timepicker/pipes/active-minute.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var ActiveMinutePipe = /** @class */ (function () {
    function ActiveMinutePipe() {
    }
    ActiveMinutePipe.prototype.transform = function (minute, currentMinute, gap, isClockFaceDisabled) {
        if (minute == null || isClockFaceDisabled) {
            return false;
        }
        var defaultGap = 5;
        return ((currentMinute === minute) && (minute % (gap || defaultGap) === 0));
    };
ActiveMinutePipe.ɵfac = function ActiveMinutePipe_Factory(t) { return new (t || ActiveMinutePipe)(); };
ActiveMinutePipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "activeMinute", type: ActiveMinutePipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ActiveMinutePipe, [{
        type: Pipe,
        args: [{
                name: 'activeMinute'
            }]
    }], function () { return []; }, null); })();
    return ActiveMinutePipe;
}());
export { ActiveMinutePipe };
